superchain/superchain.go

```diff
@@ -171,105 +172,170 @@ func (c *ChainConfig) setNilHardforkTimestampsToDefaultOrZero(s *SuperchainConfi
 	}
 }
 
-// EnhanceYAML creates a customized yaml string from a RollupConfig. After completion,
-// the *yaml.Node pointer can be used with a yaml encoder to write the custom format to file
-func (c *ChainConfig) EnhanceYAML(ctx context.Context, node *yaml.Node) error {
-	hexStringRegex := regexp.MustCompile(`^0x[a-fA-F0-9]+$`)
-
-	// Check if context is done before processing
-	if err := ctx.Err(); err != nil {
-		return fmt.Errorf("context error: %w", err)
+func (c ChainConfig) MarshalTOML() ([]byte, error) {
+	// Uses []outField to deterministically set the order of the toml based on the order of fields
+	// in the ChainConfig struct. Otherwise the fields are ordered alphabetically
+	type outField struct {
+		key   string
+		value interface{}
 	}
-
-	if node.Kind == yaml.DocumentNode && len(node.Content) > 0 {
-		node = node.Content[0] // Dive into the document node
+	var out []outField
+	v := reflect.ValueOf(c)
+
+	processTag := func(tag string) string {
+		if tag == "-" {
+			return ""
+		} else if tag != "" {
+			key := strings.Split(string(tag), ",")
+			return key[0]
+		} else {
+			return ""
+		}
 	}
 
-	var lastKey string
-	for i := 0; i < len(node.Content)-1; i += 2 {
-		keyNode := node.Content[i]
-		valNode := node.Content[i+1]
-		if valNode.Kind == yaml.ScalarNode && valNode.Tag == "!!str" && hexStringRegex.MatchString(valNode.Value) {
-			valNode.Style = yaml.DoubleQuotedStyle
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldName := v.Type().Field(i).Name
+		fieldTag, _ := v.Type().Field(i).Tag.Lookup("toml")
+		if fieldName == "HardForkConfiguration" {
+			hardForkConfig := field.Interface().(HardForkConfiguration)
+			hardForkVal := reflect.ValueOf(hardForkConfig)
+
+			for j := 0; j < hardForkVal.NumField(); j++ {
+				hfField := hardForkVal.Field(j)
+				hfFieldName := hardForkVal.Type().Field(j).Name
+				hfFieldTag, _ := hardForkVal.Type().Field(j).Tag.Lookup("toml")
+
+				if hfFieldTag == "" {
+					hfFieldTag = hfFieldName
+				}
+
+				if hfField.IsNil() {
+					continue
+				}
+
+				tag := processTag(hfFieldTag)
+				out = append(out, outField{tag, *hfField.Interface().(*uint64)})
+			}
+		} else if fieldName == "Addresses" {
+			nested, err := field.Interface().(AddressList).MarshalTOML()
+			if err != nil {
+				return nil, err
+			}
+			nestedMap := make(map[string]interface{})
+			if err := toml.Unmarshal(nested, &nestedMap); err != nil {
+				return nil, err
+			}
```

Comment:
Why is this necessary (marshalling and then unmarshalling into a different type?

Reply:
I needed to create a generic `map[string]interface{}` version of the `AddressList` in order to append it to the `out []outField` slice, which is what is used to preserve the order of fields and the ensure correct formatting. Without this whole `else if fieldName == "Addresses"` block the formatting of the output toml file looks like this. Notice that `addresses` is not formatted as a proper toml tag:
```
        	            	+addresses = AddressManager = "0x3eb579b25F6b9547e0073c848389a768FD382296"
        	            	+AnchorStateRegistryProxy = "0x03b82AE60989863BCEb0BbD442A70568e5AefB85"
        	            	+BatchSubmitter = "0x19CC7073150D9f5888f09E0e9016d2a39667df14"
        	            	+Challenger = "0x8c20c40180751d93E939DDDee3517AE0d1EBeAd2"
```

Reply:
OK. Shall we add a comment explaining this, at least?

Reply:
https://github.com/ethereum-optimism/superchain-registry/pull/386/commits/d00a3b0588294f9711f7c75c3b913338225a4c1a